pkg/receive/hashring.go

```diff
@@ -177,6 +178,143 @@ func newKetamaHashring(endpoints []Endpoint, sectionsPerNode int, replicationFac
 	}, nil
 }
 
+// groupByAZ groups endpoints by Availability Zone and sorts them by their inferred ordinal in a k8s statefulset.
+// It returns a 2D slice where each inner slice represents an AZ (sorted alphabetically)
+// and contains endpoints sorted by ordinal. All inner slices are truncated to the
+// length of the largest common sequence of ordinals starting from 0 across all AZs.
+func groupByAZ(endpoints []Endpoint) ([][]Endpoint, error) {
+	if len(endpoints) == 0 {
+		return nil, errors.New("no endpoints provided")
+	}
+
+	// Group endpoints by AZ and then by ordinal.
+	azEndpoints := make(map[string]map[int]Endpoint)
+	for _, ep := range endpoints {
+		ordinal, err := strutil.ExtractPodOrdinal(ep.Address)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to extract ordinal from address %s", ep.Address)
+		}
+		if _, ok := azEndpoints[ep.AZ]; !ok {
+			azEndpoints[ep.AZ] = make(map[int]Endpoint)
+		}
+		if _, exists := azEndpoints[ep.AZ][ordinal]; exists {
+			return nil, fmt.Errorf("duplicate endpoint ordinal %d for address %s in AZ %s", ordinal, ep.Address, ep.AZ)
+		}
+		azEndpoints[ep.AZ][ordinal] = ep
```

Comment:
there is an assumption that ordinal always 0-indexed and no gaps from 0 to X, otherwise it could have problems.